### ISSUE TITLE
Resolving IEnumerable<ISomeInterface> resolves different than with Unity 4

### ISSUE THREAD

**Issue by astaerk** [unitycontainer/unity#260]
### Description

Resolving an IEnumerable<ISomeInterface> resolves to a list with one instance instead of an empty list.
Using Unity 4 this resolves to an empty list, as expected

### To Reproduce

```C#
public interface ITestClass
{
    Guid Id { get; set; }
}

public class TestClass : ITestClass
{
    public Guid Id { get; set; }
}

[TestMethod]
public void ListOfInterfacesShouldResolveEmpty()
{
    IUnityContainer container = new UnityContainer();


    List<InjectionMember> injectionMembers = new List<InjectionMember>();

    injectionMembers.Add(new InjectionConstructor(new object[] { }));

    container.RegisterType(typeof(IEnumerable<>), typeof(List<>), null, new TransientLifetimeManager(), injectionMembers.ToArray());


    container.RegisterType(typeof(ITestClass), typeof(TestClass), null, new TransientLifetimeManager());


    var res = container.Resolve<IEnumerable<ITestClass>>();

    Assert.AreEqual(0, res.Count());
}
```

**Additional context**

In case it is of any help: I can upload a Visual Studio solution with 2 console applications. These applications share the same code files, but one uses Unity 4 and one uses Unity 5.
Using Unity 4 the resolved list is empty, using Unity 5 the list contains one item.

**Comment by ENikS** [comment]
This behavior is by design. Unity 4 did not have IEnumerable<type> supports so it used type[] in background. Unity 5 introduced natively resolving IEnumerable. The more in depth info could be found [here](https://github.com/unitycontainer/unity/wiki/Difference-between-ResolveAll-and-Enumerable.)

**Comment by ENikS** [comment]
You could resolve it as ``container.Resolve<ITestClass[]>();``

**Comment by astaerk** [comment]
> You don't need this...
Yeah, I thought so. But thanks for the hint. I will remove this from my code. I needed this with Unity 4.

Well, unfortunately it is not that easy to change, because I'm in a library and I can't really influence the uesd types. The type being resolved is determined in my code by analyzing the expected return type of a given method (expression tree). If the method wants to return an IEnumerable<Something> my library is trying to provide that.
This is why I registered IEnumerable<> in Unity 4; so if a method 'requested' IEnumerable it was resolved as a List<> internally.

I guess now I have to detect when someone wants an IEnumerable and then internally make a new generic type with IList<Something> instead of IEnumerable<Something> before requesting from Unity. I did a proof of concept with this and it seems to work.

Thanks for your help/time.
